mars/metric/backends/metric.py

```diff
@@ -0,0 +1,163 @@
+    """A float protected by a mutex."""
+
+    def __init__(self):
+        self._value = 0.0
+        self._lock = Lock()
+
+    def inc(self, amount):
+        with self._lock:
+            self._value += amount
+
+    def set(self, value):
+        with self._lock:
+            self._value = value
+
+    def get(self):
+        with self._lock:
+            return self._value
+
+
+class Metric(ABC):
+    """Base class of metrics."""
+
+    _type = None
+
+    def __init__(
+        self, name: str, description: str = "", tag_keys: Optional[Tuple[str]] = None
+    ):
+        self._name = name
+        self._description = description
+        self._tag_keys = tuple(tag_keys) if tag_keys else tuple()
+        self._init()
+
+    @property
+    def type(self):
+        return self._type
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def description(self):
+        return self._description
+
+    @property
+    def tag_keys(self):
+        return self._tag_keys
+
+    @abstractmethod
+    def _init(self):
+        """Some initialization in subclass."""
+        pass
+
+    def record(self, value=1, tags: Optional[Dict[str, str]] = None):
+        """A public method called by users."""
+        pass
+
+    @abstractmethod
+    def _record(self, value: float = 1.0, tags: Optional[Dict[str, str]] = None):
+        """An internal method called by record() and should be
+        implemented by different metric backends.
+        """
+        pass
+
+
+class Counter(Metric):
+    """A counter records the counts of events."""
+
+    _type = "counter"
+
+    def __init__(
+        self, name: str, description: str = "", tag_keys: Optional[Tuple[str]] = None
+    ):
+        super().__init__(name, description, tag_keys)
+        self._count = _MutexValue()
+
+    def record(self, value=1, tags: Optional[Dict[str, str]] = None):
+        self._count.inc(value)
+        self._record(self._count.get(), tags)
+
+
+class Gauge(Metric):
+    """A gauge represents a single numerical value that can be
+    arbitrarily set.
+    """
+
+    _type = "gauge"
+
+    def record(self, value=1, tags: Optional[Dict[str, str]] = None):
+        self._record(value, tags)
+
+
+class Meter(Metric):
+    """A meter measures the rate at which a set of events occur."""
+
+    _type = "meter"
+
+    def __init__(
+        self, name: str, description: str = "", tag_keys: Optional[Tuple[str]] = None
+    ):
+        super().__init__(name, description, tag_keys)
+        self._count = _MutexValue()
+        self._last_time = time.time()
+
+    def record(self, value=1, tags: Optional[Dict[str, str]] = None):
+        self._count.inc(value)
+        now = time.time()
+        past = now - self._last_time
+        if self._count.get() >= _THRESHOLD or past >= _RECORDED_INTERVAL_SECS:
+            qps = self._count.get() / past
+            self._record(qps, tags)
+            self._last_time = now
+            self._count.set(0)
```

Comment:
Though `self._count` is mutex value, but here, `set` and previous `inc`, can you guarantee it's threadsafe?

Reply:
Thanks for your catch. There is the same thread safe problem in `Counter`, `Meter` and `Histogram` `record()` method.  Adding lock to `record` can guarantee thread safe but it will add extra consumption and reduce performance. Maybe we can implement a non thread safe version because there are few scenarios where metrics are used in multiple threads of mars. What do you think?

Reply:
yeah, that's what I mean, we don't consider primitive here, just consider it in the service.

Reply:
Ok. Then `_MutexValue` can be removed.

Reply:
We can add a doc that this module is not threadsafe.

Reply:
Ok.